<a href="https://colab.research.google.com/github/vssood/WU_DL/blob/master/WU_DL_WK3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 3.2: Introduction to Tensorflow and Keras

In [41]:
import tensorflow as tf
import numpy as np


In [42]:
# 1x2 matrix
matrix1 = tf.constant([[3., 3.]])
# 2x1 matrix
matrix2 = tf.constant([[2.], [2.]])

product = tf.matmul(matrix1, matrix2)
print(product)
print(float(product))

tf.Tensor([[12.]], shape=(1, 1), dtype=float32)
12.0


### Subtract

In [43]:
x = tf.Variable([1.0, 2.0])
a = tf.constant([3.0, 3.0])

sub = tf.subtract(x, a)
print(sub)
print(sub.numpy())

tf.Tensor([-2. -1.], shape=(2,), dtype=float32)
[-2. -1.]


Assign / change varaible value

In [44]:
x.assign([4.0, 6.0])

<tf.Variable 'UnreadVariable' shape=(2,) dtype=float32, numpy=array([4., 6.], dtype=float32)>

In [45]:
sub = tf.subtract(x, a)
print(sub)
print(sub.numpy())

tf.Tensor([1. 3.], shape=(2,), dtype=float32)
[1. 3.]


### Keras

#### Regression with MPG input dataset

In [46]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation
import pandas as pd
import io
import os 
import requests
import numpy as np
from sklearn import metrics

In [47]:
file_path = "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv"

In [48]:
df = pd.read_csv(file_path, na_values=['NA', '?'])

In [49]:
df.head()

mpg  cylinders  displacement  ...  year  origin                       name
0  18.0          8         307.0  ...    70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...    70       1          buick skylark 320
2  18.0          8         318.0  ...    70       1         plymouth satellite
3  16.0          8         304.0  ...    70       1              amc rebel sst
4  17.0          8         302.0  ...    70       1                ford torino

[5 rows x 9 columns]

In [50]:
cars = df['name']

In [51]:
df.head(1)

mpg  cylinders  displacement  ...  year  origin                       name
0  18.0          8         307.0  ...    70       1  chevrolet chevelle malibu

[1 rows x 9 columns]

Check missing values

In [52]:
df.shape

(398, 9)

In [53]:
df.isna().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
year            0
origin          0
name            0
dtype: int64

In [54]:
# Handle missing value
df['horsepower'] = df['horsepower'].fillna(df['horsepower'].median())

In [55]:
df.columns.values

array(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin', 'name'], dtype=object)

In [56]:
# Pandas to Numpy
x = df[['cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin']].values
y = df['mpg'].values    # regression

In [57]:
# Build Neural Network
model = Sequential()
model.add(Dense(25, input_dim = x.shape[1], activation='relu'))   # hidden 1
model.add(Dense(10, activation='relu'))                 # Hidden 2
model.add(Dense(1))         # output 
model.compile(loss='mean_squared_error', optimizer='adam')
model.fit(x, y, verbose=2, epochs=100)


Epoch 1/100
13/13 - 0s - loss: 226699.7188
Epoch 2/100
13/13 - 0s - loss: 87207.8594
Epoch 3/100
13/13 - 0s - loss: 42639.2812
Epoch 4/100
13/13 - 0s - loss: 23439.2188
Epoch 5/100
13/13 - 0s - loss: 10800.5645
Epoch 6/100
13/13 - 0s - loss: 3948.7974
Epoch 7/100
13/13 - 0s - loss: 1049.8458
Epoch 8/100
13/13 - 0s - loss: 310.8604
Epoch 9/100
13/13 - 0s - loss: 234.4959
Epoch 10/100
13/13 - 0s - loss: 239.5697
Epoch 11/100
13/13 - 0s - loss: 235.5028
Epoch 12/100
13/13 - 0s - loss: 232.2986
Epoch 13/100
13/13 - 0s - loss: 231.3969
Epoch 14/100
13/13 - 0s - loss: 231.6414
Epoch 15/100
13/13 - 0s - loss: 231.0472
Epoch 16/100
13/13 - 0s - loss: 230.8707
Epoch 17/100
13/13 - 0s - loss: 230.8278
Epoch 18/100
13/13 - 0s - loss: 230.4282
Epoch 19/100
13/13 - 0s - loss: 230.1508
Epoch 20/100
13/13 - 0s - loss: 229.7843
Epoch 21/100
13/13 - 0s - loss: 229.6536
Epoch 22/100
13/13 - 0s - loss: 229.1895
Epoch 23/100
13/13 - 0s - loss: 229.1347
Epoch 24/100
13/13 - 0s - loss: 228.6751
Epoch 25/100

### Neural Network Hyperparameters

Prediction

In [58]:
pred = model.predict(x)
print(f"Shape:{pred.shape}")
print(pred[0:10])

Shape:(398, 1)
[[23.45822 ]
 [24.35388 ]
 [22.54991 ]
 [22.712797]
 [23.012663]
 [29.060377]
 [28.679228]
 [28.558043]
 [29.268538]
 [25.147642]]


Measure the peformance of the prediction

In [59]:
# Measure RMSE error.  RMSE is common for regression.
score = np.sqrt(metrics.mean_squared_error(pred,y))
print(f"Final Score (RMSE): {score}")

Final Score (RMSE): 13.787785041463286


In [60]:
# Sample prediction  - Print first 10 predictions 
for i in range(10): 
    print(f"{i+1}. Car Name: {cars[i]}, MPG: {y[i]}, "
    + f"Predicted MPG: {pred[i]}")

1. Car Name: chevrolet chevelle malibu, MPG: 18.0, Predicted MPG: [23.45822]
2. Car Name: buick skylark 320, MPG: 15.0, Predicted MPG: [24.35388]
3. Car Name: plymouth satellite, MPG: 18.0, Predicted MPG: [22.54991]
4. Car Name: amc rebel sst, MPG: 16.0, Predicted MPG: [22.712797]
5. Car Name: ford torino, MPG: 17.0, Predicted MPG: [23.012663]
6. Car Name: ford galaxie 500, MPG: 15.0, Predicted MPG: [29.060377]
7. Car Name: chevrolet impala, MPG: 14.0, Predicted MPG: [28.679228]
8. Car Name: plymouth fury iii, MPG: 14.0, Predicted MPG: [28.558043]
9. Car Name: pontiac catalina, MPG: 14.0, Predicted MPG: [29.268538]
10. Car Name: amc ambassador dpl, MPG: 15.0, Predicted MPG: [25.147642]


###  Classification: Iris

In [61]:
from tensorflow.keras.callbacks import EarlyStopping

In [62]:
file_path = "https://data.heatonresearch.com/data/t81-558/iris.csv"

In [65]:
df_iris = pd.read_csv(file_path, na_values=["NA", "?"])

In [66]:
df_iris.head()

sepal_l  sepal_w  petal_l  petal_w      species
0      5.1      3.5      1.4      0.2  Iris-setosa
1      4.9      3.0      1.4      0.2  Iris-setosa
2      4.7      3.2      1.3      0.2  Iris-setosa
3      4.6      3.1      1.5      0.2  Iris-setosa
4      5.0      3.6      1.4      0.2  Iris-setosa

Convert to nympy - classification

In [68]:
df_iris.columns.values

array(['sepal_l', 'sepal_w', 'petal_l', 'petal_w', 'species'],
      dtype=object)

In [70]:
x = df_iris[['sepal_l', 'sepal_w', 'petal_l', 'petal_w']].values
dummies = pd.get_dummies(df_iris['species'])            # classification
species = dummies.columns
y = dummies.values

In [76]:
x.shape

(150, 4)

In [72]:
dummies[0:2]

Iris-setosa  Iris-versicolor  Iris-virginica
0            1                0               0
1            1                0               0

In [73]:
species

Index(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype='object')

In [75]:
y[0:5]

array([[1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0]], dtype=uint8)

In [81]:
# Build neural network
model = Sequential()
model.add(Dense(50, input_dim = x.shape[1], activation= 'relu' ))   # Hidden 1
model.add(Dense(25, activation='relu'))                 # Hidden 2
model.add(Dense(y.shape[1], activation= 'softmax'))     # output

model.compile(loss='categorical_crossentropy', optimizer='adam')
model.fit(x, y, verbose=0, epochs=100)

In [80]:
# Print out number of species found:
print(species)

Index(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype='object')


Prediction 

Notice that three values come back for each of the 150 iris flowers

In [83]:
pred = model.predict(x)
print(f"Shape : {pred.shape}")
print(pred[0:10])

Shape : (150, 3)
[[9.9644488e-01 3.5551381e-03 1.0858597e-08]
 [9.9118340e-01 8.8164732e-03 7.5958425e-08]
 [9.9435270e-01 5.6473524e-03 4.5446804e-08]
 [9.9119240e-01 8.8075316e-03 1.2340251e-07]
 [9.9683088e-01 3.1691198e-03 1.0182261e-08]
 [9.9740607e-01 2.5938943e-03 6.0095511e-09]
 [9.9511659e-01 4.8833843e-03 4.9477212e-08]
 [9.9525297e-01 4.7470988e-03 2.2339734e-08]
 [9.8873490e-01 1.1264860e-02 2.5655649e-07]
 [9.9208128e-01 7.9187118e-03 5.4118672e-08]]


In [84]:
# Turn off the scientific notation 
np.set_printoptions(suppress=True)

In [85]:
print(y[0:10])

[[1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]]


In [87]:
# convert the predictions to the expected iris species
# The argmax function finds the index of the maximum prediction for each row.

predict_class = np.argmax(pred, axis=1)
expected_class = np.argmax(y, axis =1)
print(f"Predictions : {predict_class}")
print(f"Expected : {expected_class}")

Predictions : [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 1
 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
Expected : [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [88]:
# to to show the species name
print(species[predict_class[0:10]])

Index(['Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa'],
      dtype='object')


In [89]:
# find accuracy
correct = metrics.accuracy_score(expected_class, predict_class)
print(f"Accuracy : {correct}")

Accuracy : 0.98


### Part 3.3: Saving and Loading a Keras Neural Network

#### Save MPG model

In [92]:
save_path = '.'

In [98]:
# Save neural network structureto JSON (no weight)
model_json = model.to_json()
with open(os.path.join(save_path, "iris_netwrk.json"), "w") as json_file:
    json_file.write(model_json)

In [99]:
# save neural network structure to YAML (no weights)
model_ymal = model.to_yaml()
with open(os.path.join(save_path, 'iris_network.ymal'), "w") as ymal_file:
    ymal_file.write(model_ymal)

In [100]:
# save entire network to HDF5 (save everything, suggested)
model.save(os.path.join(save_path, "iris_network.h5"))

To load the saved model and run the prediction 

In [101]:
from tensorflow.keras.models import load_model

model2 = load_model(os.path.join(save_path, "iris_network.h5"))

In [102]:
pred = model2.predict(x)

In [ ]:
# Measure the accuracy - Cross Entropy for 

In [104]:
predict_class = np.argmax(pred, axis=1)
expected_class = np.argmax(y, axis =1)

# find accuracy
correct = metrics.accuracy_score(expected_class, predict_class)
print(f"Accuracy : {correct}")

Accuracy : 0.98


### Part 3.4: Early Stopping in Keras to Prevent Overfitting

**Early Stopping with Classification**

In [106]:
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import train_test_split

Using iris dataset to train & test split

In [109]:
# Split into validation and training sets
x_train, x_test, y_train, y_test = train_test_split(
        x, y, test_size = 0.25, random_state = 42 )

In [125]:
# Build neural network
model = Sequential()
model.add(Dense( 50, input_dim = x.shape[1], activation = 'relu' ) )   # Hidden
model.add(Dense(25, activation='relu'))             # Hidden 2
model.add(Dense(y.shape[1], activation='softmax'))  # output
model.compile(loss = 'categorical_crossentropy', optimizer='adam')

Early stopping 

In [122]:
monitor = EarlyStopping( monitor= 'val_loss',  min_delta=1e-3, patience=5, 
                        verbose = 1, mode= 'auto', restore_best_weights = True) 

In [127]:
# Fit the model 
model.fit(x_train, y_train, validation_data=(x_test, y_test),
          callbacks = [monitor], verbose = 2, epochs = 1000)

Epoch 1/1000
4/4 - 0s - loss: 0.6339 - val_loss: 0.5824
Epoch 2/1000
4/4 - 0s - loss: 0.6069 - val_loss: 0.5531
Epoch 3/1000
4/4 - 0s - loss: 0.5830 - val_loss: 0.5308
Epoch 4/1000
4/4 - 0s - loss: 0.5636 - val_loss: 0.5124
Epoch 5/1000
4/4 - 0s - loss: 0.5466 - val_loss: 0.4960
Epoch 6/1000
4/4 - 0s - loss: 0.5305 - val_loss: 0.4804
Epoch 7/1000
4/4 - 0s - loss: 0.5156 - val_loss: 0.4668
Epoch 8/1000
4/4 - 0s - loss: 0.5021 - val_loss: 0.4534
Epoch 9/1000
4/4 - 0s - loss: 0.4881 - val_loss: 0.4416
Epoch 10/1000
4/4 - 0s - loss: 0.4770 - val_loss: 0.4317
Epoch 11/1000
4/4 - 0s - loss: 0.4673 - val_loss: 0.4209
Epoch 12/1000
4/4 - 0s - loss: 0.4552 - val_loss: 0.4083
Epoch 13/1000
4/4 - 0s - loss: 0.4448 - val_loss: 0.3972
Epoch 14/1000
4/4 - 0s - loss: 0.4341 - val_loss: 0.3872
Epoch 15/1000
4/4 - 0s - loss: 0.4243 - val_loss: 0.3780
Epoch 16/1000
4/4 - 0s - loss: 0.4154 - val_loss: 0.3678
Epoch 17/1000
4/4 - 0s - loss: 0.4062 - val_loss: 0.3588
Epoch 18/1000
4/4 - 0s - loss: 0.3985 - 

In [129]:
# Predict Accuracy 
pred = model.predict(x_test)
predict_class = np.argmax(pred, axis = 1)
expected_class = np.argmax(y_test, axis =1)
correct = metrics.accuracy_score(expected_class, predict_class)
print(f"Accuracy : {correct}")

Accuracy : 1.0


**Early Stopping with Regression**

In [131]:
df.head(2)

mpg  cylinders  displacement  ...  year  origin                       name
0  18.0          8         307.0  ...    70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...    70       1          buick skylark 320

[2 rows x 9 columns]

In [132]:
cars = df['name']

In [133]:
df.columns.values

array(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin', 'name'], dtype=object)

In [136]:
x = df[['cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin']]
y = df['mpg']

In [137]:
# split test and train 
x_train_mpg, x_test_mpg, y_train_mpg, y_test_mpg = train_test_split(
    x, y, test_size = 0.25, random_state = 42 )


In [148]:
# Build Neural Network 

model = Sequential()
model.add(Dense(25, input_dim = x.shape[1], activation='relu'))  # Hidden 1
model.add(Dense(10, activation='relu'))     # Hidden 2
model.add(Dense(1))         # output
model.compile(loss = 'mean_squared_error', optimizer='adam')

In [149]:
# Early stopping

monitor = EarlyStopping(monitor = 'val_loss', min_delta= 1e-3,  
                        patience = 5, verbose = 1, mode = 'auto',
                        restore_best_weights = True)

In [150]:
# fit the model 
model.fit(x_train_mpg, y_train_mpg, validation_data=(x_test_mpg, y_test_mpg),
          callbacks  = [monitor], verbose =  0, epochs = 1000)

Restoring model weights from the end of the best epoch.
Epoch 00122: early stopping


In [151]:
# Evaluate the error
pred = model.predict(x_test_mpg)
score = np.sqrt(metrics.mean_squared_error(pred, y_test_mpg))
print(f"Final Score (RMSE) : {score}")

Final Score (RMSE) : 6.237766436141136


### Manual Neural Network Calculation

In [161]:
# Create a dataset for the XOR function
x = np.array([
              [0, 0],
              [1, 0],
              [0, 1],
              [1, 1]
])

y = np.array([
              0,
              1,
              1,
              0
])

In [162]:
print(x, y)

[[0 0]
 [1 0]
 [0 1]
 [1 1]] [0 1 1 0]


In [164]:
# Build the network
# sgd = optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)

done = False
cycle = 1

while not done:
    print("Cyle # {}".format(cycle))
    cycle +=1
    model = Sequential()
    model.add(Dense(2, input_dim = 2, activation='relu'))
    model.add(Dense(1))
    model.compile(loss = 'mean_squared_error', optimizer='adam')
    model.fit(x,y, verbose = 0, epochs=1000)

    # predict 
    pred = model.predict(x)

    # check if sucessful
# done = pred[0]<0.01 and pred[3] <0.01 and pred[1] > 0.9 and pred[2] > 0.9
    done = pred[0]<0.01 and pred[3]<0.01 and pred[1] > 0.9 \
        and pred[2] > 0.9 
    print(pred)

Cyle # 1
[[0.5252784 ]
 [0.5252469 ]
 [0.5252784 ]
 [0.24206212]]
Cyle # 2
[[0.43732506]
 [0.5045229 ]
 [0.47092167]
 [0.5381195 ]]
Cyle # 3
[[0.48646683]
 [0.48646683]
 [0.48646683]
 [0.48646683]]
Cyle # 4
[[0.00810108]
 [0.99229985]
 [0.99831873]
 [0.00814819]]
